File: src/com/edlplan/osu/support/slider/Draw3DLinePath.java
package com.edlplan.osu.support.slider;

import com.edlplan.andengine.Triangle3DBuilder;
import com.edlplan.framework.math.FMath;
import com.edlplan.framework.math.Vec2;
import com.edlplan.framework.math.line.AbstractPath;

public class Draw3DLinePath {
    private static final int MAXRES = 24;

    // Precomputed rotation constants for arc caps (same trick as DrawLinePath)
    private static final float CAP_STEP = FMath.Pi / MAXRES;
    private static final float CAP_STEP_COS = (float) Math.cos(CAP_STEP);
    private static final float CAP_STEP_SIN = (float) Math.sin(CAP_STEP);

    public float alpha;
    public float width;

    private float zEdge = -1, zCenter = 1;

    // Precomputed path structure — width-independent, computed once per path
    private float[] segTheta;
    private float[] segNormX;
    private float[] segNormY;
    private int structureSize;

    private Triangle3DBuilder triangles;
    private AbstractPath path;

    public Draw3DLinePath(AbstractPath p, float width, float zCenter, float zEdge) {
        this.zCenter = zCenter;
        this.zEdge = zEdge;
        alpha = 1;
        this.width = width;
        prepareForPath(p);
    }

    public Draw3DLinePath() {
        alpha = 1;
    }

    public void setZCenter(float zCenter) { this.zCenter = zCenter; }
    public void setZEdge(float zEdge) { this.zEdge = zEdge; }

    // -------------------------------------------------------------------------
    // Two-phase API (mirrors DrawLinePath)
    // -------------------------------------------------------------------------

    public Draw3DLinePath prepareForPath(AbstractPath p) {
        this.path = p;
        int n = p.size();
        if (n >= 2) {
            int segs = n - 1;
            if (segTheta == null || segTheta.length < segs) {
                segTheta = new float[segs];
                segNormX = new float[segs];
                segNormY = new float[segs];
            }
            for (int i = 0; i < segs; i++) {
                Vec2 a = p.get(i);
                Vec2 b = p.get(i + 1);
                segTheta[i] = Vec2.calTheta(a, b);
                float dx = b.x - a.x;
                float dy = b.y - a.y;
                float len = (float) Math.sqrt(dx * dx + dy * dy);
                if (len > 1e-6f) {
                    segNormX[i] = -dy / len;
                    segNormY[i] = dx / len;
                } else {
                    segNormX[i] = 0f;
                    segNormY[i] = 0f;
                }
            }
        }
        structureSize = n;
        return this;
    }

    public Triangle3DBuilder buildForWidth(float width, float zCenter, float zEdge,
                                           Triangle3DBuilder builder) {
        this.width = width;
        this.zCenter = zCenter;
        this.zEdge = zEdge;
        builder.reset();
        triangles = builder;
        init();
        return builder;
    }

    // -------------------------------------------------------------------------
    // Legacy single-use API (kept for applyToScene)
    // -------------------------------------------------------------------------

    public Triangle3DBuilder getTriangles() {
        if (triangles == null) {
            triangles = new Triangle3DBuilder();
        } else {
            triangles.reset();
        }
        init();
        return triangles;
    }

    // -------------------------------------------------------------------------
    // Geometry generation — all floats, zero allocation
    // -------------------------------------------------------------------------

    private void addLineCap(Vec2 org, float theta, float thetaDiff) {
        float dir = Math.signum(thetaDiff);
        if (dir == 0f) return;
        thetaDiff *= dir;
        int amountPoints = (int) Math.ceil(thetaDiff / CAP_STEP);
        if (amountPoints == 0) return;

        if (dir < 0) theta += FMath.Pi;

        float ux = (float) Math.cos(theta);
        float uy = (float) Math.sin(theta);
        float prevX = ux * width + org.x;
        float prevY = uy * width + org.y;

        final float cs = CAP_STEP_COS;
        final float ss = dir * CAP_STEP_SIN;

        float orgX = org.x, orgY = org.y;

        for (int i = 1; i <= amountPoints; i++) {
            float ux2, uy2;
            if (i == amountPoints && i * CAP_STEP > thetaDiff) {
                float finalAngle = theta + dir * thetaDiff;
                ux2 = (float) Math.cos(finalAngle);
                uy2 = (float) Math.sin(finalAngle);
            } else {
                ux2 = ux * cs - uy * ss;
                uy2 = uy * cs + ux * ss;
            }
            float x2 = ux2 * width + org.x;
            float y2 = uy2 * width + org.y;

            triangles.add(
                    orgX, orgY, zCenter,
                    prevX, prevY, zEdge,
                    x2, y2, zEdge
            );

            prevX = x2; prevY = y2;
            ux = ux2; uy = uy2;
        }
    }

    private void addLineQuads(int segIdx, Vec2 ps, Vec2 pe) {
        float nx = segNormX[segIdx] * width;
        float ny = segNormY[segIdx] * width;

        float slx = ps.x + nx, sly = ps.y + ny;
        float srx = ps.x - nx, sry = ps.y - ny;
        float elx = pe.x + nx, ely = pe.y + ny;
        float erx = pe.x - nx, ery = pe.y - ny;
        float psx = ps.x, psy = ps.y;
        float pex = pe.x, pey = pe.y;

        triangles.add(psx, psy, zCenter, pex, pey, zCenter, elx, ely, zEdge);
        triangles.add(psx, psy, zCenter, elx, ely, zEdge,   slx, sly, zEdge);
        triangles.add(psx, psy, zCenter, erx, ery, zEdge,   pex, pey, zCenter);
        triangles.add(psx, psy, zCenter, srx, sry, zEdge,   erx, ery, zEdge);
    }

    private void init() {
        int n = structureSize;
        if (n < 2) {
            if (n == 1) {
                addLineCap(path.get(0), FMath.Pi, FMath.Pi);
                addLineCap(path.get(0), 0, FMath.Pi);
            }
            return;
        }

        float theta = segTheta[0];
        addLineCap(path.get(0), theta + FMath.PiHalf, FMath.Pi);
        addLineQuads(0, path.get(0), path.get(1));

        if (n == 2) {
            addLineCap(path.get(1), theta - FMath.PiHalf, FMath.Pi);
            return;
        }

        float preTheta = theta;
        for (int i = 1; i < n - 1; i++) {
            float nextTheta = segTheta[i];
            addLineCap(path.get(i), preTheta - FMath.PiHalf, nextTheta - preTheta);
            addLineQuads(i, path.get(i), path.get(i + 1));
            preTheta = nextTheta;
        }
        addLineCap(path.get(n - 1), preTheta - FMath.PiHalf, FMath.Pi);
    }
}

File: src/com/edlplan/osu/support/slider/Draw3DLinePath.kt
package com.edlplan.osu.support.slider

import com.edlplan.andengine.Triangle3DBuilder
import com.edlplan.framework.math.FMath
import com.edlplan.framework.math.Vec2
import com.edlplan.framework.math.line.AbstractPath

class Draw3DLinePath {

    var alpha: Float = 0f
    var width: Float = 0f

    private var zEdge: Float = -1f
    private var zCenter: Float = 1f

    private var segTheta: FloatArray? = null
    private var segNormX: FloatArray? = null
    private var segNormY: FloatArray? = null
    private var structureSize: Int = 0

    internal var triangles: Triangle3DBuilder? = null
    private var path: AbstractPath? = null

    constructor(p: AbstractPath, width: Float, zCenter: Float, zEdge: Float) {
        this.zCenter = zCenter
        this.zEdge = zEdge
        alpha = 1f
        this.width = width
        prepareForPath(p)
    }

    constructor() {
        alpha = 1f
    }

    fun setZCenter(zCenter: Float) {
        this.zCenter = zCenter
    }

    fun setZEdge(zEdge: Float) {
        this.zEdge = zEdge
    }

    fun prepareForPath(p: AbstractPath): Draw3DLinePath {
        this.path = p
        val n = p.size()
        if (n >= 2) {
            val segs = n - 1
            if (segTheta == null || segTheta!!.size < segs) {
                segTheta = FloatArray(segs)
                segNormX = FloatArray(segs)
                segNormY = FloatArray(segs)
            }
            var a = p[0]
            for (i in 0 until segs) {
                val b = p[i + 1]
                segTheta!![i] = Vec2.calTheta(a, b)
                val dx = b.x - a.x
                val dy = b.y - a.y
                val len = Math.sqrt((dx * dx + dy * dy).toDouble()).toFloat()
                if (len > 1e-6f) {
                    segNormX!![i] = -dy / len
                    segNormY!![i] = dx / len
                } else {
                    segNormX!![i] = 0f
                    segNormY!![i] = 0f
                }
                a = b
            }
        }
        structureSize = n
        return this
    }

    fun buildForWidth(
        width: Float, zCenter: Float, zEdge: Float,
        builder: Triangle3DBuilder
    ): Triangle3DBuilder {
        this.width = width
        this.zCenter = zCenter
        this.zEdge = zEdge
        builder.reset()
        triangles = builder
        init()
        return builder
    }

    fun getTriangles(): Triangle3DBuilder {
        val t = triangles
        return if (t == null) {
            val builder = Triangle3DBuilder()
            triangles = builder
            init()
            builder
        } else {
            t.reset()
            init()
            t
        }
    }

    private fun addLineCap(org: Vec2, theta: Float, thetaDiff: Float) {
        val dir = Math.signum(thetaDiff)
        if (dir == 0f) return
        var thetaVar = theta
        val thetaDiffVar = thetaDiff * dir
        val amountPoints = Math.ceil((thetaDiffVar / CAP_STEP).toDouble()).toInt()
        if (amountPoints == 0) return

        if (dir < 0) thetaVar += FMath.Pi

        var ux = Math.cos(thetaVar.toDouble()).toFloat()
        var uy = Math.sin(thetaVar.toDouble()).toFloat()
        var prevX = ux * width + org.x
        var prevY = uy * width + org.y

        val cs = CAP_STEP_COS
        val ss = dir * CAP_STEP_SIN

        val orgX = org.x
        val orgY = org.y

        for (i in 1..amountPoints) {
            var ux2: Float
            var uy2: Float
            if (i == amountPoints && i * CAP_STEP > thetaDiffVar) {
                val finalAngle = thetaVar + dir * thetaDiffVar
                ux2 = Math.cos(finalAngle.toDouble()).toFloat()
                uy2 = Math.sin(finalAngle.toDouble()).toFloat()
            } else {
                ux2 = ux * cs - uy * ss
                uy2 = uy * cs + ux * ss
            }
            val x2 = ux2 * width + org.x
            val y2 = uy2 * width + org.y

            triangles!!.add(
                orgX, orgY, zCenter,
                prevX, prevY, zEdge,
                x2, y2, zEdge
            )

            prevX = x2
            prevY = y2
            ux = ux2
            uy = uy2
        }
    }

    private fun addLineQuads(segIdx: Int, ps: Vec2, pe: Vec2) {
        val nx = segNormX!![segIdx] * width
        val ny = segNormY!![segIdx] * width

        val slx = ps.x + nx
        val sly = ps.y + ny
        val srx = ps.x - nx
        val sry = ps.y - ny
        val elx = pe.x + nx
        val ely = pe.y + ny
        val erx = pe.x - nx
        val ery = pe.y - ny
        val psx = ps.x
        val psy = ps.y
        val pex = pe.x
        val pey = pe.y

        triangles!!.add(psx, psy, zCenter, pex, pey, zCenter, elx, ely, zEdge)
        triangles!!.add(psx, psy, zCenter, elx, ely, zEdge, slx, sly, zEdge)
        triangles!!.add(psx, psy, zCenter, erx, ery, zEdge, pex, pey, zCenter)
        triangles!!.add(psx, psy, zCenter, srx, sry, zEdge, erx, ery, zEdge)
    }

    private fun init() {
        val n = structureSize
        if (n < 2) {
            if (n == 1) {
                val p0 = path!![0]
                addLineCap(p0, FMath.Pi, FMath.Pi)
                addLineCap(p0, 0f, FMath.Pi)
            }
            return
        }

        var prev = path!![0]
        var next = path!![1]
        val theta = segTheta!![0]
        addLineCap(prev, theta + FMath.PiHalf, FMath.Pi)
        addLineQuads(0, prev, next)

        if (n == 2) {
            addLineCap(next, theta - FMath.PiHalf, FMath.Pi)
            return
        }

        var preTheta = theta
        for (i in 1 until n - 1) {
            prev = next
            next = path!![i + 1]
            val nextTheta = segTheta!![i]
            addLineCap(prev, preTheta - FMath.PiHalf, nextTheta - preTheta)
            addLineQuads(i, prev, next)
            preTheta = nextTheta
        }
        addLineCap(next, preTheta - FMath.PiHalf, FMath.Pi)
    }

    companion object {
        private const val MAXRES = 24
        private val CAP_STEP = FMath.Pi / MAXRES
        private val CAP_STEP_COS = Math.cos(CAP_STEP.toDouble()).toFloat()
        private val CAP_STEP_SIN = Math.sin(CAP_STEP.toDouble()).toFloat()
    }
}
